File: src/main/java/net/flyingfishflash/loremlist/domain/lrmlistitem/LrmListItemService.java
package net.flyingfishflash.loremlist.domain.lrmlistitem;

import java.util.List;
import java.util.UUID;
import kotlin.Pair;
import kotlin.Triple;
import net.flyingfishflash.loremlist.domain.ServiceResponse;
import net.flyingfishflash.loremlist.domain.lrmitem.data.LrmItemCreate;
import net.flyingfishflash.loremlist.domain.lrmlistitem.data.LrmListItemAdded;

public interface LrmListItemService {
  ServiceResponse<Long> countByOwnerAndListId(UUID listId, String owner);

  ServiceResponse<LrmListItemAdded> add(UUID id, List<UUID> idCollection, String componentsOwner);

  ServiceResponse<LrmListItem> create(UUID listId, LrmItemCreate lrmItemCreate, String creator);

  ServiceResponse<LrmListItem> findByOwnerAndItemIdAndListId(
      UUID itemId, UUID listId, String owner);

  ServiceResponse<Triple<String, String, String>> move(
      UUID itemId, UUID currentListId, UUID destinationListId, String owner);

  void patchQuantity(LrmListItem patchedLrmListItem);

  void patchIsSuppressed(LrmListItem patchedLrmListItem);

  ServiceResponse<Pair<String, String>> removeByOwnerAndListIdAndItemId(
      UUID listId, UUID itemId, String owner);

  ServiceResponse<Pair<String, Integer>> removeByOwnerAndItemId(UUID itemId, String owner);

  ServiceResponse<Pair<String, Integer>> removeByOwnerAndListId(UUID listId, String owner);

  // item context
  ServiceResponse<Long> countByOwnerAndItemId(UUID itemId, String owner);

  void patchName(LrmListItem patchedLrmListItem);

  void patchDescription(LrmListItem patchedLrmListItem);
}

File: src/main/kotlin/net/flyingfishflash/loremlist/domain/lrmlistitem/LrmListItemService.kt
package net.flyingfishflash.loremlist.domain.lrmlistitem

import net.flyingfishflash.loremlist.domain.ServiceResponse
import net.flyingfishflash.loremlist.domain.lrmitem.data.LrmItemCreate
import net.flyingfishflash.loremlist.domain.lrmlistitem.data.LrmListItemAdded
import java.util.UUID

interface LrmListItemService {
  fun countByOwnerAndListId(listId: UUID, owner: String): ServiceResponse<Long>
  fun add(id: UUID, idCollection: List<UUID>, componentsOwner: String): ServiceResponse<LrmListItemAdded>
  fun create(listId: UUID, lrmItemCreate: LrmItemCreate, creator: String): ServiceResponse<LrmListItem>
  fun findByOwnerAndItemIdAndListId(itemId: UUID, listId: UUID, owner: String): ServiceResponse<LrmListItem>
  fun move(itemId: UUID, currentListId: UUID, destinationListId: UUID, owner: String): ServiceResponse<Triple<String, String, String>>
  fun patchQuantity(patchedLrmListItem: LrmListItem)
  fun patchIsSuppressed(patchedLrmListItem: LrmListItem)
  fun removeByOwnerAndListIdAndItemId(listId: UUID, itemId: UUID, owner: String): ServiceResponse<Pair<String, String>>
  fun removeByOwnerAndItemId(itemId: UUID, owner: String): ServiceResponse<Pair<String, Int>>
  fun removeByOwnerAndListId(listId: UUID, owner: String): ServiceResponse<Pair<String, Int>>

  // item context
  fun countByOwnerAndItemId(itemId: UUID, owner: String): ServiceResponse<Long>
  fun patchName(patchedLrmListItem: LrmListItem)
  fun patchDescription(patchedLrmListItem: LrmListItem)
}
